Pass `rm` as self-removal function

christianbundy opened the pull request:
This binds the `ready()` function so that `this()` actually runs `rm()` on itself, that way listeners have a way to remove themselves without any JavaScript gymnastics. This is technically a breaking change because we're changing `this`, so maybe if/when this is merged we could release as obv@1.0.0?

## Before

```js
var rm = sv.since(function (v) {
  if(v === log.since.value) {
    var _cb = cb; cb = null; _cb()
  }
  if(!cb && rm) rm()
})
```

## After

```js
sv.since(function (v) {
  if(v === log.since.value) {
    var _cb = cb; cb = null; _cb()
    this()
  }
})
```

---

Edited the above examples because the original had slightly different behavior as it didn't protect against the double-`cb()` race condition.

christianbundy committed:
Pass `this` as self-removal function

dominictarr commented:
this looks like a good suggestion. but would you consider it to be a breaking change if it was `function (v, rm)` instead of `this()` to remove?

dominictarr commented:
but I think there a problem. if you register two listeners and the first one removes itself, does the second one still get called:

it iterates over the array, incrementally, so if you remove something it steps over the gap?
https://github.com/dominictarr/obv/blob/master/index.js#L7-L10

christianbundy committed:
Pass `rm()` as function parameter, add more tests

Instead of passing the self-destruct function as `this`, this commit
changes the behavior so that it's passed as a secondary function
parameter.

This also adds a test to ensure that if you register two
listeners and the first one removes itself, the second one still gets
called without any problems.

> if you register two listeners and the first one removes itself, does the second one still get called:

christianbundy commented:
Thanks for the review! I like that pattern much more, I've added a commit so that `rm` is passed as a function argument rather than binding to `this`. I've also increased the test coverage so that a listener removing itself doesn't break other listeners.

dominictarr committed:
failing test cases, when a listener is removed while listeners are being triggered

dominictarr commented:
@christianbundy I found some failing cases, see a9cbc10c8c189bf1e3b9daabd3bda01183eee6a8
I just copied your test cases, and tried different variations of removing listeners and keeping them.

christianbundy commented:
Thanks! I'll take a look and see if I can reason about how to resolve that.

dominictarr commented:
[a hint](https://github.com/dominictarr/obv/pull/2#issuecomment-464500518)

christianbundy committed:
Remove listener by setting null instead of splicing

christianbundy commented:
Thanks! I've replaced that `splice()` call with an operation that sets that list item to `null`, which circumvents the step-over issue. Please let me know if you notice anything else!

dominictarr commented:
that has correct behaviour now, but I'm a bit worried that this makes the list grow over time - if there were things set and removed lots of listeners you'd see a slow down, because it would have to step over the null entries.

can you think of another way to do it?

christianbundy committed:
Clone array and splice original to remove listener

christianbundy commented:
Oof. I thought that if we copied the array then we'd have an issue where it could get double-triggered, but I didn't realize that `isUnchanged()` is actually covering that scenario. Thanks for the prod, is this closer to what you were thinking?

dominictarr commented:
I don't think that `rm` function does what you think it does. it'll wrap `i` but if you call `rm` after the whole is gone it will unsplice at the current value of `i`... that's why the old `rm` checks if `ready==listeners[i]`

https://github.com/dominictarr/obv/blob/1fa9eaca1da0a3b53e14e613bf0f80836aeb50fa/index.js#L30-L36

Copying the whole array will probably be slow, given `obv` is in the hot path in flume stuff I don't want to make it any slower. Already, I wrote https://github.com/dominictarr/flumelog-aligned-offset using less observables and it was much faster.

I was trying to think of different ways to do it, like keeping track of the index, and how many things have been removed... I think either setting `null` then cleaning the array (only if something was removed) would work... but maybe the best way would be to use a doubly linked list instead of an array - then you can remove items while iterating with no problems.

christianbundy commented:
Is there a JS library you like for doubly linked lists, or is this in the hot path enough that we need to roll our own implementation?

dominictarr commented:
If performance is the goal, we should start with a benchmark, and also maybe look at how unsubscribe is actually used in the code. linked lists look simple but are easy to get wrong! (I've done it)
then compare the current method, a couple of libraries, and _then_ try your own implementation.
ideally, you start a benchmark race that the other authors get into, and they improve on the libraries and beat your design, then you have the fastest possible code.

I had some success with that appoach here for example: https://github.com/dominictarr/bench-lru
